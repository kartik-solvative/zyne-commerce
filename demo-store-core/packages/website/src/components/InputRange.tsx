import { useEffect, useLayoutEffect, useRef, useState } from 'react'
import styles from './InputRange.module.scss'

type Props = {
  defaultValue?: readonly [number, number]
  min?: number
  max?: number
  step?: number
  onRelease?: (value: number[]) => void
  format?: (value: number) => string
}

export const InputRange: React.FC<Props> = ({ min = 0, max = 100, defaultValue = [min, max], step = 1, onRelease = () => {}, format = value => value.toString() }) => {
  const progressRef = useRef<HTMLDivElement>(null)
  const minLabelRef = useRef<HTMLDivElement>(null)
  const minInputRef = useRef<HTMLInputElement>(null)
  const maxLabelRef = useRef<HTMLDivElement>(null)

  const [defaultMinValue, defaultMaxValue] = defaultValue

  const [minValue, setMinValue] = useState<number>(defaultMinValue < min ? min : defaultMinValue)
  const [maxValue, setMaxValue] = useState<number>(defaultMaxValue > max ? max : defaultMaxValue)

  const gap = 0

  useLayoutEffect(function handleUI() {
    const leftPercentage = (((minValue - min) / (max - min)) * 100)
    const rightPercentage = (100 -((maxValue - min) / (max - min)) * 100)
    const styleLeft = `calc(${leftPercentage}% - ${16 / 100 * leftPercentage}px)`
    const styleRight = `calc(${rightPercentage}% - ${16 / 100 * rightPercentage}px)`

    if (progressRef.current) {
      progressRef.current.style.left = styleLeft
      progressRef.current.style.right = styleRight
    }

    if (minLabelRef.current) {
      minLabelRef.current.style.left = styleLeft
    }

    if (maxLabelRef.current) {
      maxLabelRef.current.style.right = styleRight
    } 

    if (minInputRef.current) {
      if (minValue === max) {
        minInputRef.current.style.zIndex = '1'
      } else {
        minInputRef.current.style.zIndex = ''
      }
    }
  }, [minValue, maxValue, min, max])

  useEffect(function handleDefaultValueChange() {
    setMinValue(defaultMinValue < min ? min : defaultMinValue)
    setMaxValue(defaultMaxValue > max ? max : defaultMaxValue)
  }, [defaultMinValue, defaultMaxValue, min, max])

  const handleMinValueChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    const minVal = event.currentTarget.valueAsNumber
    const maxVal = maxValue

    if ((maxVal - minVal) < gap) {
      setMinValue(maxVal - gap)
    } else {
      setMinValue(minVal)
    }
  }

  const handleMaxValueChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    const minVal = minValue
    const maxVal = event.currentTarget.valueAsNumber

    if ((maxVal - minVal) < gap) {
      setMaxValue(minVal + gap)
    } else {
      setMaxValue(maxVal)
    }
  }

  const handleRelease = () => {
    onRelease([minValue, maxValue])
  }

  return (
    <div>
      <div data-testid='number-input-container' className={styles.numberInput}>
        <div className='field'>
          <span>Min</span>
          <input data-testid='number-input-1' onChange={handleMinValueChange} type='number' min={min} value={minValue} step={step} />
        </div>
        <div className='separator'>-</div>
        <div className='field'>
          <span>Max</span>
          <input data-testid='number-input-2' onChange={handleMaxValueChange} type='number' max={max} value={maxValue} step={step} />
        </div>
      </div>
      <div className={styles.slider}>
        <div data-testid='progress' className='progress' ref={progressRef}></div>
      </div>
      <div data-testid='range-input' className={styles.rangeInput}>
        <div aria-hidden={true} ref={minLabelRef} className='text-gray-400 text-xs top-3 absolute left-0'>{format(minValue)}</div>
        <input className='cursor-grab' ref={minInputRef} data-testid='range-input-1' onChange={handleMinValueChange} onMouseUp={handleRelease} onTouchEnd={handleRelease} onKeyUp={handleRelease} type='range' min={min} max={max} value={minValue} step={step} />
        <div aria-hidden={true} ref={maxLabelRef} className='text-gray-400 text-xs top-3 absolute right-0'>{format(maxValue)}</div>
        <input className='cursor-grab' data-testid='range-input-2' onChange={handleMaxValueChange} onMouseUp={handleRelease} onTouchEnd={handleRelease} onKeyUp={handleRelease} type='range' min={min} max={max} value={maxValue} step={step} />
      </div>
    </div>
  )
}