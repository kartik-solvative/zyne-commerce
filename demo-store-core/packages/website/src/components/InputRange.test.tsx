import { fireEvent, render } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { InputRange } from './InputRange'


test('should match the snapshot', () => {
  const { container } = render(<InputRange />)

  expect(container).toMatchSnapshot()
})

test('should render with default properties', () => {
  const { getByTestId } = render(<InputRange />)

  expect(getByTestId('number-input-1')).toHaveValue(0)
  expect(getByTestId('number-input-2')).toHaveValue(100)
  expect(getByTestId('range-input-1')).toHaveValue('0')
  expect(getByTestId('range-input-2')).toHaveValue('100')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(0% - 0px)',
    right: 'calc(0% - 0px)'
  })
})

test('should render changing the default values', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} />)

  expect(getByTestId('number-input-1')).toHaveValue(10)
  expect(getByTestId('number-input-2')).toHaveValue(90)
  expect(getByTestId('range-input-1')).toHaveValue('10')
  expect(getByTestId('range-input-2')).toHaveValue('90')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(10% - 1.6px)',
    right: 'calc(10% - 1.6px)'
  })
})

test('should change the lower defaultValue when "min" is set', () => {
  const { getByTestId } = render(<InputRange min={-10} />)

  expect(getByTestId('number-input-1')).toHaveValue(-10)
  expect(getByTestId('number-input-2')).toHaveValue(100)
  expect(getByTestId('range-input-1')).toHaveValue('-10')
  expect(getByTestId('range-input-2')).toHaveValue('100')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(0% - 0px)',
    right: 'calc(0% - 0px)'
  })
})

test('should change the upper defaultValue when "max" is set', () => {
  const { getByTestId } = render(<InputRange max={250} />)

  expect(getByTestId('number-input-1')).toHaveValue(0)
  expect(getByTestId('number-input-2')).toHaveValue(250)
  expect(getByTestId('range-input-1')).toHaveValue('0')
  expect(getByTestId('range-input-2')).toHaveValue('250')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(0% - 0px)',
    right: 'calc(0% - 0px)'
  })
})

test('should change the provided value if lower than "min"', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} min={50} />)

  expect(getByTestId('number-input-1')).toHaveValue(50)
  expect(getByTestId('number-input-2')).toHaveValue(90)
  expect(getByTestId('range-input-1')).toHaveValue('50')
  expect(getByTestId('range-input-2')).toHaveValue('90')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(0% - 0px)',
    right: 'calc(20% - 3.2px)'
  })
})

test('should change the provided value if greater than "max"', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} max={50} />)

  expect(getByTestId('number-input-1')).toHaveValue(10)
  expect(getByTestId('number-input-2')).toHaveValue(50)
  expect(getByTestId('range-input-1')).toHaveValue('10')
  expect(getByTestId('range-input-2')).toHaveValue('50')
  expect(getByTestId('progress')).toHaveStyle({
    left: 'calc(20% - 3.2px)',
    right: 'calc(0% - 0px)'
  })
})

test('should change the value when user slide without violating min-max rules', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 90]} min={10} max={90} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '0' } })
  })

  expect(rangeInput1).toHaveValue('10')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '50' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '100' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('90')

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '80' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('80')
})

test('lower value cannot be greater than upper value', () => {
  const { getByTestId } = render(<InputRange defaultValue={[10, 50]} min={10} max={90} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '60' } })
  })

  expect(rangeInput1).toHaveValue('50')
  expect(rangeInput2).toHaveValue('50')

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '20' } })
  })

  act(() => {
    fireEvent.change(rangeInput2, { target: { value: '10' } })
  })

  expect(rangeInput1).toHaveValue('20')
  expect(rangeInput2).toHaveValue('20')
})

test('should render the progress bar correctly', () => {
  const { getByTestId } = render(<InputRange min={9} max={205} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')
  const progress = getByTestId('progress')

  expect(rangeInput1).toHaveValue('9')
  expect(rangeInput2).toHaveValue('205')
  expect(progress).toHaveStyle({
    left: 'calc(0% - 0px)',
    right: 'calc(0% - 0px)'
  })
})

test('should set zIndex to 1 when minValue match the max value', () => {
  const { getByTestId } = render(<InputRange min={10} max={90} />)

  const rangeInput1 = getByTestId('range-input-1')

  expect(rangeInput1).not.toHaveStyle({
    zIndex: '1'
  })

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '90' } })
  })

  expect(rangeInput1).toHaveStyle({
    zIndex: '1'
  })

  act(() => {
    fireEvent.change(rangeInput1, { target: { value: '89' } })
  })

  expect(rangeInput1).not.toHaveStyle({
    zIndex: '1'
  })
})

test('should trigger onRelease when range changes and mouseUp event is triggered', () => {
  const onRelease = jest.fn()
  const { getByTestId } = render(<InputRange min={10} max={90} onRelease={onRelease} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.mouseDown(rangeInput1)
    fireEvent.change(rangeInput1, { target: { value: '50' } })
  })

  expect(onRelease).not.toBeCalled()

  act(() => {
    fireEvent.mouseUp(rangeInput1)
  })

  act(() => {
    fireEvent.mouseDown(rangeInput2)
    fireEvent.change(rangeInput2, { target: { value: '80' } })
  })

  act(() => {
    fireEvent.mouseUp(rangeInput2)
  })

  expect(onRelease).toBeCalledTimes(2)
  expect(onRelease.mock.calls[0][0]).toStrictEqual([50, 90])
  expect(onRelease.mock.calls[1][0]).toStrictEqual([50, 80])
})

test('should trigger onRelease when range changes and keyUp event is triggered', () => {
  const onRelease = jest.fn()
  const { getByTestId } = render(<InputRange min={10} max={90} onRelease={onRelease} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.keyDown(rangeInput1)
    fireEvent.change(rangeInput1, { target: { value: '20' } })
  })

  expect(onRelease).not.toBeCalled()

  act(() => {
    fireEvent.keyUp(rangeInput1)
  })

  act(() => {
    fireEvent.keyDown(rangeInput2)
    fireEvent.change(rangeInput2, { target: { value: '76' } })
  })

  act(() => {
    fireEvent.keyUp(rangeInput2)
  })

  expect(onRelease).toBeCalledTimes(2)
  expect(onRelease.mock.calls[0][0]).toStrictEqual([20, 90])
  expect(onRelease.mock.calls[1][0]).toStrictEqual([20, 76])
})

test('should trigger onRelease when range changes and touchEnd event is triggered', () => {
  const onRelease = jest.fn()
  const { getByTestId } = render(<InputRange min={10} max={90} onRelease={onRelease} />)

  const rangeInput1 = getByTestId('range-input-1')
  const rangeInput2 = getByTestId('range-input-2')

  act(() => {
    fireEvent.touchStart(rangeInput1)
    fireEvent.change(rangeInput1, { target: { value: '34' } })
  })

  expect(onRelease).not.toBeCalled()

  act(() => {
    fireEvent.touchEnd(rangeInput1)
  })

  act(() => {
    fireEvent.touchStart(rangeInput2)
    fireEvent.change(rangeInput2, { target: { value: '71' } })
  })

  act(() => {
    fireEvent.touchEnd(rangeInput2)
  })

  expect(onRelease).toBeCalledTimes(2)
  expect(onRelease.mock.calls[0][0]).toStrictEqual([34, 90])
  expect(onRelease.mock.calls[1][0]).toStrictEqual([34, 71])
})
