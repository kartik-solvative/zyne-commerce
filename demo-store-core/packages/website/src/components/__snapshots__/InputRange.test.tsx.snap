// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div>
  <div>
    <div
      class="numberInput"
      data-testid="number-input-container"
    >
      <div
        class="field"
      >
        <span>
          Min
        </span>
        <input
          data-testid="number-input-1"
          min="0"
          step="1"
          type="number"
          value="0"
        />
      </div>
      <div
        class="separator"
      >
        -
      </div>
      <div
        class="field"
      >
        <span>
          Max
        </span>
        <input
          data-testid="number-input-2"
          max="100"
          step="1"
          type="number"
          value="100"
        />
      </div>
    </div>
    <div
      class="slider"
    >
      <div
        class="progress"
        data-testid="progress"
        style="left: calc(0% - 0px); right: calc(0% - 0px);"
      />
    </div>
    <div
      class="rangeInput"
      data-testid="range-input"
    >
      <div
        aria-hidden="true"
        class="text-gray-400 text-xs top-3 absolute left-0"
        style="left: calc(0% - 0px);"
      >
        0
      </div>
      <input
        class="cursor-grab"
        data-testid="range-input-1"
        max="100"
        min="0"
        step="1"
        type="range"
        value="0"
      />
      <div
        aria-hidden="true"
        class="text-gray-400 text-xs top-3 absolute right-0"
        style="right: calc(0% - 0px);"
      >
        100
      </div>
      <input
        class="cursor-grab"
        data-testid="range-input-2"
        max="100"
        min="0"
        step="1"
        type="range"
        value="100"
      />
    </div>
  </div>
</div>
`;
